### ISSUE TITLE
Rule proposal: no-wildcard-import

### ISSUE THREAD

**Issue by silvenon** [eslint/eslint#4865]
Found [this rule](https://github.com/airbnb/javascript#10.2) in Airbnb's JS styleguide and I liked it.

**Comment by eslintbot** [comment]
@silvenon Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by silvenon** [comment]
1. When does this rule warn? Please describe and show example code:

 It warns when you use a wildcard import.

 ```js
 import * as foo from 'foo';
 ```

2. Is this rule preventing an error or is it stylistic?

 Stylistic.

3. Why is this rule a candidate for inclusion instead of creating a custom rule?

 Not sure what this question means :confused: 

4. Are you willing to create the rule yourself?

 Yes.

**Comment by nzakas** [comment]
>Why is this rule a candidate for inclusion instead of creating a custom rule?

 In other words, why should this be part of the ESLint core and managed by us when you can write and manage it yourself?

**Comment by silvenon** [comment]
After giving it some more thought, I don't think it's a good rule to enforce. Forcing a single default export instead of named exports provides no benefit.

**Comment by michaelficarra** [comment]
That syntax doesn't import the default export, it imports *all* exports.

**Comment by silvenon** [comment]
Yes. I meant that not allowing wildcard imports will force you to write modules with a single (default) export.

**Comment by michaelficarra** [comment]
No it won't.

**Comment by silvenon** [comment]
```js
// foo.js
export const a = 5;
export const b = 6;
```

```js
// bar.js
import * as foo from 'foo'; // oops, error, fixing...
import foo from 'foo'; // but that won't work, gotta rewrite foo.js...
```

```js
// foo.js
export default {
  a: 5,
  b: 6
};
```

```js
import foo from 'foo'; // this works now
console.log(foo.a);
```

**Comment by silvenon** [comment]
Oh, I see what you meant, when you're importing only certain exports, it doesn't really matter whether it's a single exported object or a named export.

**Comment by michaelficarra** [comment]
Yes.

```js
import {a, b} from "foo";
```

**Comment by jackson-sandland** [comment]
What about the case when you are importing actions into a component? e.g. import * as awesomeActions from './actions/awesomeActions';

**Comment by csvan** [comment]
> After giving it some more thought, I don't think it's a good rule to enforce. 

I disagree. Stylistic benefits aside, disallowing wildcard imports has several technical implications. For example, wildcard imports breaks treeshaking in Webpack.

**Comment by ljharb** [comment]
Can this be reopened? I'd very much like to ban wildcard imports in all of my projects.

**Comment by platinumazure** [comment]
Can this be enforced with `no-restricted-syntax`?

**Comment by j-f1** [comment]
> Can this be enforced with `no-restricted-syntax`?

Yes — just restrict `ImportNamespaceSpecifier`.

**Comment by silvenon** [comment]
I hadn't thought about that, good idea!

```js
{
  "no-restricted-syntax": [
    "error",
    {
      "selector": ":matches(ImportNamespaceSpecifier, ExportAllDeclaration, ExportNamespaceSpecifier)",
      "message": "Import/export only modules you need"
    }
  ]
}
```

I haven't tested this, but it should prohibit all wildcard import and export statements:

```js
import * as foo from 'foo';
export * from 'foo';
export * as foo from 'foo';
```

If this works for you, @ljharb, I can close this.

**Comment by ljharb** [comment]
It can, but that’s a blunt instrument that works very poorly with shared configs - since anyone wishing to modify any of the config for that rule needs to copy-paste the entire config and then tweak, to do it properly.

I think `import *` syntax is special enough, and will warrant special options like “don’t warn if you’re reflecting on the ModuleRecord object”, that it needs its own rule.

**Comment by silvenon** [comment]
Also, people will still often want to do `import * as React from 'react'` since that is recommended in Flow docs. Yeah, that's a huge tradeoff… I wish `no-restricted-syntax` is a special rule that can be stacked.

I agree that wildcard import is enough of an antipattern to warrant its own rule. How would configuration for this new rule look like? What would whitelisting look like?

**Comment by not-an-aardvark** [comment]
> It can, but that’s a blunt instrument that works very poorly with shared configs - since anyone wishing to modify any of the config for that rule needs to copy-paste the entire config and then tweak, to do it properly.

I agree. I think it would be a good idea to improve this situation (e.g. with https://github.com/eslint/eslint/issues/9192) so that `no-restricted-syntax` can be improved for all cases.

> I think `import *` syntax is special enough, and will warrant special options like “don’t warn if you’re reflecting on the ModuleRecord object”, that it needs its own rule.

I disagree, this rule seems completely redundant with `no-restricted-syntax` to me. If `no-restricted-syntax` is inconvenient to use, then I think we should work on fixing that directly rather than adding a bunch of redundant rules in the meantime. The point of `no-restricted-syntax` is to make rules like this unnecessary.

**Comment by ljharb** [comment]
I don’t think it makes them unnecessary; however, it should be able to make their *implementation* simply defer to that of no-restricted-syntax. There’s not actually much value in minimizing the number of distinct rules; whereas there is much value in abstracting away complex configuration (especially anything that deals with AST nodes) from eslint users.

**Comment by ilyavolodin** [comment]
> There’s not actually much value in minimizing the number of distinct rules; whereas there is much value in abstracting away complex configuration
Completely disagree. ESLint has a huge number of rules. It's incredibly hard to find a rule that enforces specific pattern. Even as one of the maintainer, I have to constantly dig through documentation whenever creating my own config from the scratch. There is a very substantial value in minimizing the number of distinct rules from my perspective.

**Comment by Chaoste** [comment]
I wouldn't completely disagree. Imagine you would have only 10 eslint rules which are capable of replacing any other rule - I can't imagine that everyone is able to understand the logic/language you would need to implement for this. I see no reason why making things complicated and screw up this nice module.

If a programmer installs eslint and looks for this feature we won't say "Hey great. I have to learn so much stuff before I can start !". Vice versa imagine we would google for this rule and find it directly. In my opinion a lint library shouldn't be minimalistic. I agree that it should provide advanced rules for advanced users with very special rules. But it should also abstract away complexity by providing simple rules. 

Btw: if you're always starting your config from scratch you're definitely doing something wrong...

**Comment by Chaoste** [comment]
For those who are wondering how to add this rule and add an exception for Reacts wildcard import:
```yaml
no-restricted-syntax:
    - 2
    - selector: :matches(ImportNamespaceSpecifier[local.name!='React']) # just add another square bracket for a second name matching
      message: Import/export only modules you need
```

Took me some time to make this work since the documentation on selectors is missing a lot of configurations (like the parameters of ImportNamespaceSpecifier). If you want to add similar rules or would like to see what options are available beside the documentation:
1. [List of available selectors [Code]](https://github.com/eslint/eslint/blob/2e68be643178eeb86f2b9f66bc1670b624cb09f2/lib/rules/indent.js#L23)

2. [Some code snippets from babel-generator making clear which parameters are available.](http://www.bkbits.net/u/toish/chromatism/node_modules/babel-generator/lib/generators/modules.js?PAGE=anno&REV=573534eaCSgy_FY4H2jaMnpNd3wpSA)

Greetings,
Thomas

**Comment by ljharb** [comment]
(Not sure why you’d need an exception for React; `import React from ‘react’` is the proper way to import it)

**Comment by michaelficarra** [comment]
@Chaoste See https://github.com/estree/estree for the tree documentation.

**Comment by Chaoste** [comment]
Thanks for the feedback! If it's not necessary I would also remove the exception.

And thanks for pointing out the background of AST selectors - didn't know what this is all about ;)
